### ISSUE TITLE
suggest Ok-wrapping for `?`-originating type mismatch when (and only when) applicable

### ISSUE THREAD

**Issue by zackmdavis** [rust-lang/rust#55429]
Sometimes E0308 on a try-expression can be fixed by wrapping with `Ok`. This is a somewhat common mistake in the form of a `?`-expression appearing in tail position of a function that is expected to `Result`—

```rust
fn foo() -> Result<isize, ()> {
     Ok(1)
}
 
fn bar() -> Result<isize, ()> {
    foo()? // `Ok(foo()?)` would work
}

fn main() {}
```

However, not all type mismatches on `?` are like this. Here's a negative example—

```rust
fn maybe_numbers() -> Result<Vec<i32>, ()> {
    Ok(vec![1, 2, 3])
}

fn try_it() -> Result<String, ()> {
    let n: i32 = maybe_numbers()?; // `Ok(maybe_numbers()?)` is not the right move here
    Ok(format!("{:?}", n))
}

fn main() {}
```

We want to suggest `Ok`-wrapping when that's the correct fix. (This issue is being filed because a [previous attempt](https://github.com/rust-lang/rust/issues/51632) at this suggestion had too many false positives and is [likely to be removed](https://github.com/rust-lang/rust/pull/55423).)

See [discussion on #52537 for](https://github.com/rust-lang/rust/issues/52537#issuecomment-433124819) one idea on how to proceed.

**Comment by estebank** [comment]
Current output for the first case:

```
error[E0308]: try expression alternatives have incompatible types
 --> src/main.rs:6:5
  |
6 |     foo()?
  |     ^^^^^-
  |     |    |
  |     |    help: try removing this `?`
  |     expected enum `std::result::Result`, found isize
  |
  = note: expected type `std::result::Result<isize, ()>`
             found type `isize`
```

**Comment by estebank** [comment]
Current output:
```
error[[E0308]](https://doc.rust-lang.org/stable/error_codes/E0308.html): `?` operator has incompatible types
 --> src/main.rs:6:5
  |
6 |     foo()?
  |     ^^^^^^ expected `Result<isize, ()>`, found `isize`
  |
  = note: `?` operator cannot convert from `isize` to `Result<isize, ()>`
  = note: expected enum `Result<isize, ()>`
             found type `isize`
help: try removing this `?`
  |
6 -     foo()?
6 +     foo()
  |
help: try wrapping the expression in `Ok`
  |
6 |     Ok(foo()?)
  |     +++      +
```
```
error[[E0308]](https://doc.rust-lang.org/stable/error_codes/E0308.html): `?` operator has incompatible types
 --> src/main.rs:6:18
  |
6 |     let n: i32 = maybe_numbers()?; // `Ok(maybe_numbers()?)` is not the right move here
  |                  ^^^^^^^^^^^^^^^^ expected `i32`, found `Vec<i32>`
  |
  = note: `?` operator cannot convert from `Vec<i32>` to `i32`
  = note: expected type `i32`
           found struct `Vec<i32>`
```
